static/js/tools/stat_var/dataset_selector.tsx

```diff
@@ -0,0 +1,231 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Component to select a source and dataset.
+ */
+
+import axios from "axios";
+import React, { useEffect, useState } from "react";
+import { Card, Container, CustomInput } from "reactstrap";
+
+import { NamedPlace } from "../../shared/types";
+
+const PREFIX = "dataset-selector";
+const MAX_SUGGESTIONS = 5;
+
+interface DatasetSelectorProps {
+  // Filter StatVarHierarchy to stat vars that exist for the entities.
+  filterStatVars: (entities: NamedPlace[]) => void;
+  // Map of source name to dcid.
+  sourceMap: Record<string, string>;
+}
+
+function simplify(s: string): string {
+  return s.replace(/[^0-9a-z]/gi, "").toLowerCase();
+}
+
+export function DatasetSelector(props: DatasetSelectorProps): JSX.Element {
+  const [activeSuggestion, setActiveSuggestion] = useState(0);
+  const [datasets, setDatasets] = useState([]);
+  const [filteredSuggestions, setFilteredSuggestions] = useState(
+    Object.keys(props.sourceMap)
+  );
+  const [showSuggestions, setShowSuggestions] = useState(false);
+  const [suggestionsWidth, setSuggestionsWidth] = useState(null);
+  const [userInput, setUserInput] = useState("");
+
+  const handleResize = () => {
+    const width = document.getElementById(`${PREFIX}-ac`).offsetWidth;
+    if (width !== suggestionsWidth) {
+      setSuggestionsWidth(width);
+    }
+  };
+
+  useEffect(() => {
+    handleResize();
+    window.addEventListener("resize", handleResize);
+    return () => {
+      window.removeEventListener("resize", handleResize);
+    };
+  });
+
+  async function updateDatasets(source: string): Promise<void> {
+    if (!(source in props.sourceMap)) {
+      return;
+    }
+    const dcid = props.sourceMap[source];
+    const datasetsPromise = await axios.get(
+      `/api/browser/propvals/isPartOf/${dcid}`
+    );
+    const currentDatasets = [];
+    for (const dataset of datasetsPromise?.data?.values?.in) {
+      currentDatasets.push({
+        dcid: dataset.dcid,
+        name: dataset.name,
+      });
+    }
+    currentDatasets.sort((a, b): number => {
+      return a.name.localeCompare(b.name);
+    });
+    setDatasets(currentDatasets);
+  }
+
+  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    handleResize();
+    const input = e.currentTarget.value;
+    let newSuggestions = Object.keys(props.sourceMap)
+      .map((s) => {
+        return {
+          name: s,
+          rank: simplify(s).indexOf(simplify(input)),
+        };
+      })
+      .sort((a, b): number => {
+        return a.rank - b.rank;
+      })
+      .filter((s) => s.rank > -1)
+      .map((s) => s.name);
+    if (newSuggestions.length > MAX_SUGGESTIONS) {
+      newSuggestions = newSuggestions.slice(0, MAX_SUGGESTIONS);
```

Comment:
I think you can just slice it without checking that length is greater than MAX_SUGGESTIONS because if length is shorter, slice will just slice up to length of the array.

Reply:
thanks!! updated